_posts/2016-08-10-dynamic-loading.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Dynamic Code Loading on a MCU
+authors:
+  - brghena
+  - alevy
+---
+
+One key feature of Tock is the ability to load and run multiple applications
+simultaneously. In a modern computer, the OS uses the Memory Management Unit
+(MMU) to provide a virtual address space for each process. The code for each
+process is written assuming that it is the only code in the world and that it
+can place memory at any address it pleases. The MMU handles translating these
+virtual addresses into a physical address in real memory, which is shared
+between all processes. Unfortunately, in the world of embedded systems MMUs are
+not available. Processors like the ARM Cortex-M series omit them because they
+aren't necessary for traditional single-process embedded devices.
+
+A common approach to handle this issue is to assign addresses to applications
+at compile-time. For example, Application A can be placed at address 0x21000
+and Application B can be placed at address 0x22000. This runs into problems
+when Application A grows in size in the future. Moreover, with Tock we want to
+be able to handle dynamically adding, updating, and removing applications at
+run time. Assigning each an address in advance simply isn't possible.
+
+In Tock, we use position independent code (PIC) [^1] to enable loading
+multiple applications. In PIC, all branches and jumps are PC-relative
+rather than absolute, allowing code to be placed at any address in
+Flash. All references to the data section are indirected through the
```

Comment:
A major motivation for the MPU was to provide a place to safely keep crypto
keys that would be accessible to the OS but not the app.

- Prabal

On Fri, Aug 12, 2016 at 12:59 PM, Pat Pannuto <notifications@github.com>
wrote:

> In _posts/2016-08-10-dynamic-loading.md
> <https://github.com/helena-project/tock-www/pull/1#discussion_r74624488>:
>
> > +virtual addresses into a physical address in real memory, which is shared
> > +between all processes. Unfortunately, in the world of embedded systems MMUs are
> > +not available. Processors like the ARM Cortex-M series omit them because they
> > +aren't necessary for traditional single-process embedded devices.
> > +
> > +A common approach to handle this issue is to assign addresses to applications
> > +at compile-time. For example, Application A can be placed at address 0x21000
> > +and Application B can be placed at address 0x22000. This runs into problems
> > +when Application A grows in size in the future. Moreover, with Tock we want to
> > +be able to handle dynamically adding, updating, and removing applications at
> > +run time. Assigning each an address in advance simply isn't possible.
> > +
> > +In Tock, we use position independent code (PIC) [^1] to enable loading
> > +multiple applications. In PIC, all branches and jumps are PC-relative
> > +rather than absolute, allowing code to be placed at any address in
> > +Flash. All references to the data section are indirected through the
>
> delete "in Flash"; confuses the point
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/helena-project/tock-www/pull/1/files/2eefbaf888b7989a6248d434f89349f325c5e0e0#r74624488>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH6xX0FD_hihSoQv_ADEvjyqxVB2YzQ1ks5qfKZ9gaJpZM4JjQwy>
> .
>